src/Models/LagrangianParticleTracking/lagrangian_particle_advection.jl

```diff
@@ -183,7 +183,7 @@ function advect_lagrangian_particles!(particles, model, Δt)
     worksize = length(particles)
 
     advect_particles_kernel! = _advect_particles!(device(arch), workgroup, worksize)
-    advect_particles_kernel!(particles.properties, particles.restitution, model.grid, Δt, total_velocities(model))
+    advect_particles_kernel!(particles.properties, model, particles.restitution, model.grid, Δt, total_velocities(model))
```

Comment:
we cannot pass the `model` to a GPU kernel because we don't have an `adapt` function for the model. 

You should extract the fields that you need and pass those to the kernel individually

Reply:
I see. Is `particles` fine? What do you think the `ParticleAdvectiveForcing` arguments should contain? I'm imagining `(particles, fluid_velocities, background_velocities, tracers, Δt)` perhaps.

Reply:
sounds good! maybe also auxiliary fields to allow customization?

Reply:
Put the same stuff in it that we put into the tendency kernel forcing functions.

Reply:
> and the buoyancy model to allow for buoyant particles

That's an interesting idea. Since no buoyancy model is grid-dependent, usually a user can include things like the equation of state as a parameter to a forcing function.

I think you should design this feature to be identical to forcing for tendencies. Literally, identical. that will make it easy to understand.

Reply:
Sorry I am slightly confused: what do you mean by buoyancy model as an argument? The way I am thinking about it whatever buoyancy model one wish to use for the lagrangian particles can be defined in the function to be passed into `ParticleAdvectionForcing`

Reply:
In line with the forcing for tendencies, which I am assuming is this:
https://github.com/CliMA/Oceananigans.jl/blob/76c32f65bbf1035686515fcc039a45cafcd8c3ff/src/Forcings/discrete_forcing.jl#L46C1-L46C1
The arguments for the advective forcing would be as such
```
u_advective_forcing(particles, p, grid, clock, Δt, model_fields)
```
Is that correct?

Reply:
we probably also need `i, j, k` as arguments

Reply:
There is no `i, j, k` though because this is particles right? The equivalent is `p`, the particle index. So the discrete form arguments would be `p, grid, clock, model_fields`.

Shouldn't you multiply this by `Δt` (because its forward Euler forcing)? I don't think you need to pass in `Δt`.

Reply:
The forcings are called here:

https://github.com/CliMA/Oceananigans.jl/blob/76c32f65bbf1035686515fcc039a45cafcd8c3ff/src/Models/NonhydrostaticModels/nonhydrostatic_tendency_kernel_functions.jl#L80

for example.

`DiscreteForcing` is a simple wrapper function that allows users to also store "parameters" in the forcing object. 

Eventually, it makes sense to use the same here, and have "discreet form" and "continuous form" for `ParticleForcing`. It's up to you whether you want to go all the way. Those wrappers are pure convenience and do not add functionality.

Reply:
> to be passed into ParticleAdvectionForcing

Should we call it simply `ParticleForcing`? It's true that this is a forcing on particle _position_ and is therefore, in some sense, "advective". But I'm not sure it actually helps clarify how the forcing is used.

Reply:
We would need `i, j, k` for `model_fields` unless we want to interpolate them and pass them

Reply:
Hmm, but what would `i, j, k` be? I guess to use `model_fields`, the user here would have to call `interpolate` manually, and they would also need to know the staggered location...

I do like the idea of interpolating the model fields, but to put that in we probably want to design an API where the interpolation only happens if the users want to use those fields, eg similar to how we have `field_dependencies` in `ContinuousForcing`. But as a first pass, we can simply avoid interpolating if `isnothing(particle_forcing)`.